### PR TITLE
IGNITE-11011 intermediate commit for discovery data processing sequen…

### PR DESCRIPTION
**Pull request by sergey-chugunov-1985**
…ce change

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDiscoDataHandlingInNewClusterTest.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        boolean mixedCachesGroupFound = false;
+
+        for (CacheGroupDescriptor grpDesc : groups.values()) {
+            if (grpDesc.cacheOrGroupName().equals(GridCacheUtils.UTILITY_CACHE_NAME))
+                defaultGroupFound = true;
+            else if (grpDesc.cacheOrGroupName().equals(MIXED_CACHES_GROUP_NAME_0))
+                mixedCachesGroupFound = true;
+        }
+
+        assertTrue(defaultGroupFound && mixedCachesGroupFound);
```
Comment (Jokser) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDiscoDataHandlingInNewClusterTest.java]: Assert message can be re-used from non-persistent test.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java [hunk 2]
```diff
@@ -135,31 +161,50 @@
...
 
     /** */
-    private void verifyCachesAndGroups(IgniteEx ignite) {
+    private void verifyCachesAndGroups(IgniteEx ignite, Collection<String> cacheNames) {
         Map<String, DynamicCacheDescriptor> caches = ignite.context().cache().cacheDescriptors();
 
-        assertEquals(2, caches.size());
-        caches.keySet().contains(GridCacheUtils.UTILITY_CACHE_NAME);
-        caches.keySet().contains(STATIC_CACHE_NAME);
+        assertEquals(3, caches.size());
```
Comment (Jokser) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java]: I think it's safer to use cacheNames.size() + 1

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java [hunk 3]
```diff
@@ -116,14 +130,26 @@
             );
         }
 
-        CacheConfiguration staticCacheCfg = new CacheConfiguration(STATIC_CACHE_NAME)
+        if (igniteInstanceName.equals(NODE_4_DIFF_STATIC_CONFIG) || igniteInstanceName.equals("clientWithDiffConfig"))
```
Comment (Jokser) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java]: Missed constant variable name usage "clientWithDiffConfig".

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 4]
```diff
@@ -4933,6 +4951,14 @@ private void processNodeAddFinishedMessage(TcpDiscoveryNodeAddFinishedMessage ms
                     mux.notifyAll();
                 }
 
+                if (gridDiscoveryData != null)
+                    spi.onExchange(gridDiscoveryData, U.resolveClassLoader(spi.ignite().configuration()));
+
+                for (DiscoveryDataPacket dataPacket : joiningNodesDiscoDataList)
```
Comment (Jokser) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: It's not clear that joiningNodesDiscoDataList will not be null in this piece of code, could you please leave a comment about it or provide explicit null check?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDiscoDataHandlingInNewClusterTest.java (hunk 1 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDiscoDataHandlingInNewClusterTest.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java (hunk 2 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java (hunk 3 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDiscoDataHandlingInNewClusterTest.java]: I introduced the constant but didn't use it here. Fixed.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java (hunk 4 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java]: Indeed there are execution branches leaving this collection uninitialized. Although I haven't seen NPE at this line anywhere in our tests it is safer to add explicit null check.

**Pull request closed without merging**